File: dev-scripts/promote-skill.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "optparse"
require "pathname"
require "set"

ALWAYS_ALLOWED_FILES = %w[
  SKILL.md
  pyproject.toml
  uv.lock
].freeze

ALWAYS_ALLOWED_DIRS = %w[
  references
  scripts
  assets
  templates
].freeze

CONDITIONAL_DIRS = %w[
  agents
  validator
  eval-viewer
].freeze

DISALLOWED_NAMES = %w[
  dev-scripts
  evals
  eval-runs
  tmp
  benchmarks
  review-output
  tests
  .venv
  .pytest_cache
  __pycache__
  justfile
  .gitignore
].freeze

DISALLOWED_PATHS = %w[
  scripts/ci
].freeze

Options = Struct.new(:replace, :dry_run, :agents_root, keyword_init: true)

def usage
  "Usage: ruby dev-scripts/promote-skill.rb <skill-name> [--replace] [--dry-run] [--agents-root PATH]"
end

def parse_options(argv)
  options = Options.new(
    replace: false,
    dry_run: false,
    agents_root: Pathname.new(".agents/skills")
  )

  parser = OptionParser.new do |opts|
    opts.banner = usage

    opts.on("--replace", "remove the existing promoted skill before copying") do
      options.replace = true
    end

    opts.on("--dry-run", "show what would be copied without changing files") do
      options.dry_run = true
    end

    opts.on("--agents-root PATH", "override promoted skills root") do |path|
      options.agents_root = Pathname.new(path)
    end
  end

  parser.parse!(argv)
  abort usage unless argv.length == 1

  [argv.first, options]
end

def repo_root
  Pathname.pwd
end

def ensure_safe_skill_name!(skill_name)
  return if skill_name.match?(/\A[a-z0-9][a-z0-9-]*\z/)

  abort "invalid skill name: #{skill_name}"
end

def skill_body(source)
  source.join("SKILL.md").read
end

def referenced?(body, name)
  body.include?("#{name}/") || body.include?("`#{name}`") || body.include?("<skill-dir>/#{name}")
end

def allowed_entries(source)
  body = skill_body(source)
  entries = []

  ALWAYS_ALLOWED_FILES.each do |file|
    entries << file if source.join(file).file?
  end

  ALWAYS_ALLOWED_DIRS.each do |dir|
    entries << dir if source.join(dir).directory?
  end

  CONDITIONAL_DIRS.each do |dir|
    next unless source.join(dir).directory?
    next unless referenced?(body, dir)

    entries << dir
  end

  entries
end

def unpromoted_entries(source, entries)
  allowed = entries.to_set
  source.children.map { |child| child.basename.to_s }.reject { |name| allowed.include?(name) }.sort
end

def copy_entry(source, destination, entry, dry_run:)
  from = source.join(entry)
  to = destination.join(entry)

  puts "copy #{from} -> #{to}"
  return if dry_run

  FileUtils.mkdir_p(to.dirname)
  from.directory? ? FileUtils.cp_r(from, to) : FileUtils.cp(from, to)
end

def disallowed_promoted_paths(destination)
  return [] unless destination.exist?

  paths = []
  destination.find do |path|
    next if path == destination

    relative = path.relative_path_from(destination).to_s
    basename = path.basename.to_s
    paths << relative if DISALLOWED_NAMES.include?(basename)
    paths << relative if DISALLOWED_PATHS.any? { |prefix| relative == prefix || relative.start_with?("#{prefix}/") }
  end
  paths.uniq.sort
end

skill_name, options = parse_options(ARGV)
ensure_safe_skill_name!(skill_name)

source = repo_root.join("skills", skill_name)
destination = repo_root.join(options.agents_root, skill_name)

abort "missing source skill: #{source}" unless source.directory?
abort "missing SKILL.md: #{source.join("SKILL.md")}" unless source.join("SKILL.md").file?

if destination.exist? && !options.replace && !options.dry_run
  abort "promoted skill already exists: #{destination} (use --replace)"
end

entries = allowed_entries(source)
abort "SKILL.md must be promoted" unless entries.include?("SKILL.md")

skipped = unpromoted_entries(source, entries)
puts "promote #{skill_name}"
puts "source: #{source}"
puts "destination: #{destination}"
puts "entries: #{entries.join(", ")}"
puts "skipped: #{skipped.join(", ")}" unless skipped.empty?

unless options.dry_run
  FileUtils.rm_rf(destination) if destination.exist? && options.replace
  FileUtils.mkdir_p(destination)
end

entries.each do |entry|
  copy_entry(source, destination, entry, dry_run: options.dry_run)
end

violations = disallowed_promoted_paths(destination)
unless violations.empty?
  abort "disallowed promoted files remain:\n#{violations.map { |path| "  - #{path}" }.join("\n")}"
end

puts options.dry_run ? "dry-run: ok" : "promote: ok"

File: dev-scripts/promote-skill.ts
#!/usr/bin/env bun

import { existsSync, mkdirSync, readdirSync, readFileSync, rmSync, cpSync, statSync } from "node:fs";
import { basename, dirname, join, relative, resolve } from "node:path";

const alwaysAllowedFiles = ["SKILL.md", "pyproject.toml", "uv.lock"];
const alwaysAllowedDirs = ["references", "scripts", "assets", "templates"];
const conditionalDirs = ["agents", "validator", "eval-viewer"];
const disallowedNames = new Set([
  "dev-scripts",
  "evals",
  "eval-runs",
  "tmp",
  "benchmarks",
  "review-output",
  "tests",
  ".venv",
  ".pytest_cache",
  "__pycache__",
  "justfile",
  ".gitignore",
]);
const disallowedPaths = ["scripts/ci"];

type Options = {
  replace: boolean;
  dryRun: boolean;
  agentsRoot: string;
};

function usage(): string {
  return "Usage: bun run dev-scripts/promote-skill.ts <skill-name> [--replace] [--dry-run] [--agents-root PATH]";
}

function fail(message: string): never {
  console.error(message);
  process.exit(1);
}

function parseOptions(argv: string[]): { skillName: string; options: Options } {
  const options: Options = {
    replace: false,
    dryRun: false,
    agentsRoot: ".agents/skills",
  };
  const positional: string[] = [];

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--replace") {
      options.replace = true;
    } else if (arg === "--dry-run") {
      options.dryRun = true;
    } else if (arg === "--agents-root") {
      const value = argv[index + 1];
      if (!value) fail(usage());
      options.agentsRoot = value;
      index += 1;
    } else if (arg.startsWith("--")) {
      fail(usage());
    } else {
      positional.push(arg);
    }
  }

  if (positional.length !== 1) fail(usage());
  return { skillName: positional[0], options };
}

function ensureSafeSkillName(skillName: string): void {
  if (/^[a-z0-9][a-z0-9-]*$/.test(skillName)) return;
  fail(`invalid skill name: ${skillName}`);
}

function isDirectory(path: string): boolean {
  return existsSync(path) && statSync(path).isDirectory();
}

function isFile(path: string): boolean {
  return existsSync(path) && statSync(path).isFile();
}

function skillBody(source: string): string {
  return readFileSync(join(source, "SKILL.md"), "utf8");
}

function referenced(body: string, name: string): boolean {
  return body.includes(`${name}/`) || body.includes(`\`${name}\``) || body.includes(`<skill-dir>/${name}`);
}

function allowedEntries(source: string): string[] {
  const body = skillBody(source);
  const entries: string[] = [];

  for (const file of alwaysAllowedFiles) {
    if (isFile(join(source, file))) entries.push(file);
  }
  for (const dir of alwaysAllowedDirs) {
    if (isDirectory(join(source, dir))) entries.push(dir);
  }
  for (const dir of conditionalDirs) {
    if (isDirectory(join(source, dir)) && referenced(body, dir)) entries.push(dir);
  }

  return entries;
}

function unpromotedEntries(source: string, entries: string[]): string[] {
  const allowed = new Set(entries);
  return readdirSync(source).filter((entry) => !allowed.has(entry)).sort();
}

function copyEntry(source: string, destination: string, entry: string, dryRun: boolean): void {
  const from = join(source, entry);
  const to = join(destination, entry);
  console.log(`copy ${from} -> ${to}`);
  if (dryRun) return;

  mkdirSync(dirname(to), { recursive: true });
  cpSync(from, to, { recursive: true });
}

function listFilesAndDirs(root: string): string[] {
  if (!existsSync(root)) return [];
  const results: string[] = [];
  const visit = (current: string): void => {
    for (const entry of readdirSync(current)) {
      const next = join(current, entry);
      results.push(next);
      if (statSync(next).isDirectory()) visit(next);
    }
  };
  visit(root);
  return results;
}

function disallowedPromotedPaths(destination: string): string[] {
  const paths: string[] = [];
  for (const path of listFilesAndDirs(destination)) {
    const rel = relative(destination, path);
    const name = basename(path);
    if (disallowedNames.has(name)) paths.push(rel);
    if (disallowedPaths.some((prefix) => rel === prefix || rel.startsWith(`${prefix}/`))) paths.push(rel);
  }
  return [...new Set(paths)].sort();
}

const { skillName, options } = parseOptions(process.argv.slice(2));
ensureSafeSkillName(skillName);

const root = process.cwd();
const source = resolve(root, "skills", skillName);
const destination = resolve(root, options.agentsRoot, skillName);

if (!isDirectory(source)) fail(`missing source skill: ${source}`);
if (!isFile(join(source, "SKILL.md"))) fail(`missing SKILL.md: ${join(source, "SKILL.md")}`);
if (existsSync(destination) && !options.replace && !options.dryRun) {
  fail(`promoted skill already exists: ${destination} (use --replace)`);
}

const entries = allowedEntries(source);
if (!entries.includes("SKILL.md")) fail("SKILL.md must be promoted");

const skipped = unpromotedEntries(source, entries);
console.log(`promote ${skillName}`);
console.log(`source: ${source}`);
console.log(`destination: ${destination}`);
console.log(`entries: ${entries.join(", ")}`);
if (skipped.length > 0) console.log(`skipped: ${skipped.join(", ")}`);

if (!options.dryRun) {
  if (existsSync(destination) && options.replace) rmSync(destination, { recursive: true, force: true });
  mkdirSync(destination, { recursive: true });
}

for (const entry of entries) copyEntry(source, destination, entry, options.dryRun);

const violations = disallowedPromotedPaths(destination);
if (violations.length > 0) {
  fail(`disallowed promoted files remain:\n${violations.map((path) => `  - ${path}`).join("\n")}`);
}

console.log(options.dryRun ? "dry-run: ok" : "promote: ok");
